Enhancement Proposal: Authentication Filter (#4235)
nginxinc/nginx-gateway-fabric: docs/proposals/authentication-filter.md

- Design for OIDC Auth
- An Auth filter for TCP and UDP routes
- Design for integration with [ExternalAuth in the Gateway API](https://gateway-api.sigs.k8s.io/geps/gep-1494/)

## Introduction

> This document focus expliclty on Authentiaction (AuthN) and not Authorization (AuthZ). Authentiaction (AuthN) defines the verification of identiy. It asks the question, "Who are you?". This is different from Authorization (AuthZ), which preceeds Authentication. It asks the question, "What are you allowed to do".

Reviewer: ```suggestion
This document focuses expliclty on Authentication (AuthN) and not Authorization (AuthZ). Authentication (AuthN) defines the verification of identiy. It asks the question, "Who are you?". This is different from Authorization (AuthZ), which preceeds Authentication. It asks the question, "What are you allowed to do".
```

type AuthenticationFilter struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

	// Spec defines the desired state of the AuthenticationFilter.
	Spec AuthenticationFilterSpec `json:"spec"`

	// Status defines the state of the AuthenticationFilter, following the same
	// pattern as SnippetsFilter: per-controller conditions with an Accepted condition.
	//
	// +optional
	Status AuthenticationFilterStatus `json:"status,omitempty"`
}

// +kubebuilder:object:root=true

// AuthenticationFilterList contains a list of AuthenticationFilter.
type AuthenticationFilterList struct {
	metav1.TypeMeta `json:",inline"`
	metav1.ListMeta `json:"metadata,omitempty"`
	Items           []AuthenticationFilter `json:"items"`
}

// AuthenticationFilterSpec defines the desired configuration.
// Exactly one of Basic or JWT must be set according to Type.
type AuthenticationFilterSpec struct {
	// Type selects the authentication mechanism.
	//
	// +kubebuilder:validation:Enum=Basic;JWT
	Type AuthType `json:"type"`

	// Basic configures HTTP Basic Authentication.
	//
	// +optional
	Basic *BasicAuth `json:"basic,omitempty"`

> 	// JWT configures JSON Web Token authentication (NGINX Plus).

Reviewer: is JWT a plus feature only?

	//
	// +optional
	JWT *JWTAuth `json:"jwt,omitempty"`
}

// AuthType defines the authentication mechanism.
type AuthType string

const (
	AuthTypeBasic AuthType = "Basic"
	AuthTypeJWT   AuthType = "JWT"
)

// BasicAuth configures HTTP Basic Authentication.
type BasicAuth struct {
	// Secret is the name of the Secret containing htpasswd data.

> 	// The Secret must be in the same namespace as this filter.

Author: This comment was mostly to reflect security constraints. I noticed I didn't make note of it elsewhere in the proposal though. Mainly, we want to make sure that the AuthenticaitonFilter can only reference secrets in the same namespace by default, and like @sjberman said we can support integration with ReferenceGrant to support cross-namespace where needed.

I'll add a section to the proposal to cover that.

> 	// The Secret must be in the same namespace as this filter.

Author: @sjberman @salonichf5 The Security Considerations section should be updated now to talk about namespace isolation as well as cross-namespace support with ReferenceGrant

> 	// The Secret must be in the same namespace as this filter.

Author: @ciarams87 I'll amend the comment around this.

> 	// The Secret must be in the same namespace as this filter.

Reviewer: I think he meant using the secretRef type as [SecretObjectReference](https://github.com/kubernetes-sigs/gateway-api/blob/82e6d271ca075c4c77b7ef467f447f1ce0b7721e/apis/v1/object_reference_types.go#L51)

	Secret string `json:"secret"`

	// Key is the key within the Secret that contains the htpasswd data.
	// Default: "htpasswd".
	//
	// +optional
	Key *string `json:"key,omitempty"`

	// Realm used by NGINX auth_basic; helps with logging and WWW-Authenticate.
	//
	// +optional
	Realm *string `json:"realm,omitempty"`

	// OnFailure customizes the 401 response for failed authentication.
	//
	// +optional
	OnFailure *AuthFailureResponse `json:"onFailure,omitempty"`
}

 // JWTAuth configures JWT-based authentication (NGINX Plus).
type JWTAuth struct {
	// Realm used by NGINX auth_jwt; sets realm in the auth challenge.
	//
	// +optional
	Realm *string `json:"realm,omitempty"`

	// Mode selects how JWT keys are provided: local file or remote JWKS.
	// Default: File.
	//
	// +optional

> 	// +kubebuilder:validation:Enum=File;Remote

Reviewer: same comment as below for specifying defaults and enum

	Mode JWTKeyMode `json:"mode,omitempty"`

	// File specifies local JWKS configuration (Secret or ConfigMap, mount path, file name).
	// Required when Mode == File. Exactly one of ConfigMapRef or SecretRef must be set.
	//
	// +optional
	File *JWTFileKeySource `json:"file,omitempty"`

	// Remote specifies remote JWKS configuration.
	// Required when Mode == Remote.
	//
	// +optional

> 	Remote *JWTRemoteKeySource `json:"remote,omitempty"`

Reviewer: I think it would be nice to have CEL rules here on how the mode is set? 

```
// +kubebuilder:validation:XValidation:rule="(self.mode == 'File') == has(self.file)",message="file must be set if mode is File"
// +kubebuilder:validation:XValidation:rule="(self.mode == 'Remote') == has(self.remote)",message="remote must be set if mode is Remote"
```

	// Example: "/etc/nginx/keys".
	MountPath string `json:"mountPath"`

	// FileName is the file name of the JWKS within the mount path.
	// Example: "jwks.json".
	FileName string `json:"fileName"`

	// KeyCache is the cache duration for keys (auth_jwt_key_cache).
	// Example: "10m".
	//
	// +optional
	KeyCache *string `json:"keyCache,omitempty"`
}

 // JWTRemoteKeySource specifies remote JWKS configuration.
type JWTRemoteKeySource struct {
	// URL is the JWKS endpoint, e.g. "https://issuer.example.com/.well-known/jwks.json".
	URL string `json:"url"`

	// Cache configures NGINX proxy_cache for JWKS fetches made via auth_jwt_key_request.
	// When set, NGF will render proxy_cache_path in http{} and attach proxy_cache to the internal JWKS location.
	//
	// +optional
	Cache *JWKSCache `json:"cache,omitempty"`
}

 // JWKSCache controls NGINX proxy_cache_path and proxy_cache settings used for JWKS responses.
type JWKSCache struct {
	// Path is the filesystem path for cached JWKS objects.
	// Example: "/var/cache/nginx/jwks".
	Path string `json:"path"`

	// Levels specifies the directory hierarchy for cached files.

> 	// Example: "1:2".

Reviewer: like 1: is for most frequently used?

	KeysZoneSize string `json:"keysZoneSize"`

	// MaxSize limits the total size of the cache (e.g., "50m").
	//
	// +optional
	MaxSize *string `json:"maxSize,omitempty"`

	// Inactive defines the inactivity timeout before cached items are evicted (e.g., "10m").
	//
	// +optional
	Inactive *string `json:"inactive,omitempty"`

	// UseTempPath controls whether a temporary file is used for cache writes.
	// Maps to use_temp_path=(on|off). Default: false (off).
	//
	// +optional
	UseTempPath *bool `json:"useTempPath,omitempty"`
}

// JWTTokenType represents NGINX auth_jwt_type.
type JWTTokenType string

const (
	JWTTokenTypeSigned    JWTTokenType = "signed"
	JWTTokenTypeEncrypted JWTTokenType = "encrypted"
	JWTTokenTypeNested    JWTTokenType = "nested"
)

// JWTRequiredClaims specifies exact-match requirements for claims.
type JWTRequiredClaims struct {
	// Issuer (iss) required exact value.
	//
	// +optional
	Iss *string `json:"iss,omitempty"`

	// Audience (aud) required exact value.
	//
	// +optional
	Aud *string `json:"aud,omitempty"`
}

// JWTTokenSource specifies where tokens may be read from.
type JWTTokenSource struct {
	// Read token from Authorization header. Default: true.
	//

> 	// +optional

Reviewer 2: If this is an optional field though, it shouldn't have a default.

> 	// +optional

Reviewer 2: I think realm was the only one we questioned needed a default in another comment. Otherwise looks good.

> 	// +optional

Author: Sounds good. I'll look over the design again and see what makes sense

> 	// +optional

Author: @sjberman @salonichf5 so as I was looking over it, nothing mandates that we provide a default value for the realm. I'm adding a section to the top of the proposal that outlines the purpose of the realm described in [RFC7617](https://www.rfc-editor.org/rfc/rfc7617).

In short, we can remove the default value and have it default to an empty string.

JWT Auth requires users to provided a bearer token through the Authroization header.

Both of these methods can be easily intercepted over HTTP.

Users that attach an `AuthenticaitonFilter` to a HTTPRoute/GRPCRoute should be advised to enable HTTPS traffic at the Gateway level for the routes.

Any exmaple configurations and deployments for the `AuthenticationFilter` should enable HTTPS at the Gateway level by default.

The `mountPath` for local JWKS should be mounted to a fixed location (e.g., /etc/nginx/keys).
The `fileName` for a local JWKS should be sanatized to a pattern of [A-Za-z0-9._-].

### Remote JWKS

Proxy cache TTL should be configurable and set to a resonable default, reducing periods of stale cached JWKs.

### Key rotation

Users sholud be advised to regularly rotate their JWKS keys in cases where they chose to reference a local JWKS via a `secrefRef` or `configMapRef`

### Auth failure behaviour

3xx response codes should not be allowed and AuthenticationFilter.onFailure must not support redirect targets. This is to prevent to prevent open-redirect abuse.

401 and 403 should be the only allowable auth failure codes.

> 

Author: These have been added now to the API

}

// JWTKeyMode selects where JWT keys come from.
// +kubebuilder:validation:Enum=File;Remote
type JWTKeyMode string

const (
  JWTKeyModeFile   JWTKeyMode = "File"
  JWTKeyModeRemote JWTKeyMode = "Remote"
)

// JWTAuth configures JWT-based authentication (NGINX Plus).
// +kubebuilder:validation:XValidation:message="mode 'File' requires file set and remote unset",rule="self.mode == 'File' ? self.file != null && self.remote == null : true"
// +kubebuilder:validation:XValidation:message="mode 'Remote' requires remote set and file unset",rule="self.mode == 'Remote' ? self.remote != null && self.file == null : true"
// +kubebuilder:validation:XValidation:message="when file is set, mode must be 'File'",rule="self.file != null ? self.mode == 'File' : true"
// +kubebuilder:validation:XValidation:message="when remote is set, mode must be 'Remote'",rule="self.remote != null ? self.mode == 'Remote' : true"
type JWTAuth struct {
  // Realm used by NGINX `auth_jwt` directive
  // https://nginx.org/en/docs/http/ngx_http_auth_jwt_module.html#auth_jwt
  // Configures "realm="<realm_value>" in WWW-Authenticate header in error page location.
  //
  // +optional

>   // +kubebuilder:default="Restricted"

Reviewer 2: Should we lowercase it then?

>   // +kubebuilder:default="Restricted"

Author: Resolving this one as the main conversation is happening in this thread here: https://github.com/nginx/nginx-gateway-fabric/pull/4235#discussion_r2565070267

  Realm *string `json:"realm,omitempty"`

  // Mode selects how JWT keys are provided: local file or remote JWKS.

>   Mode JWTKeyMode `json:"mode,omitempty"`

Reviewer 2: I think this should probably be named KeyMode to be more specific.

Reviewer 2: Sorry, when I commented this I meant the field name, not the type. The type can stay as JWTKeyMode

Reviewer 2: Same with the other fields that I left similar comments on.

  // File specifies local JWKS configuration.
  // Required when Mode == File.
  //
  // +optional
  File *JWTFileKeySource `json:"file,omitempty"`

  // Remote specifies remote JWKS configuration.
  // Required when Mode == Remote.
  //
  // +optional
  Remote *JWTRemoteKeySource `json:"remote,omitempty"`

  // Leeway is the acceptable clock skew for exp/nbf checks.
  // Configures `auth_jwt_leeway` directive.
  // https://nginx.org/en/docs/http/ngx_http_auth_jwt_module.html#auth_jwt_leeway
  // Example: "auth_jwt_leeway 60s".
  //
  // +optional

>   // +kubebuilder:default=60s

Reviewer 2: If NGINX already has a default, we don't need to define it in the policy. We can just inherit that default. When you define it in the policy like this, then we are explicitly setting it when a user creates the policy. We don't need to do that.

And if NGINX ever changes its default, we just inherit it and don't have to change our code.

>   // +kubebuilder:default=60s

Author: Thanks has been changed now

This behaviour falls in line with the expected behaviour of filters in the Gateway API, which generally allows only one type of a specific filter (authentication, rewriting, etc.) within a rule.

Below is an example of an **invalid** HTTPRoute that references multiple `AuthenticationFilter` resources in a single rule

```yaml
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name: invalid-httproute
spec:
  parentRefs:
  - name: gateway
  hostnames:
  - api.example.com
  rules:
  - matches:
    - path:
        type: PathPrefix
        value: /api
    filters:
    - type: ExtensionRef
      extensionRef:
        group: gateway.nginx.org
        kind: AuthenticationFilter
        name: basic-auth
    - type: ExtensionRef
      extensionRef:
        group: gateway.nginx.org
        kind: AuthenticationFilter
        name: jwt-auth
    backendRefs:
    - name: backend
      port: 80
```

#### Attaching a JWT AuthenticationFilter to a route when using NGINX OSS

If a user attempts to attach a JWT tpye AuthenticationFilter while using NGINX OSS, the rule referncing the filter will be `Rejected`.

> This can appear as `UnresolvedRef` to inform the user that the rule has been `Rejected`.

Reviewer 2: Is `Invalid` an option? UnresolvedRef generally means "we couldn't find this resource".

Author: @sjberman @salonichf5 let me know if you still think using `UnresolvedRef` as the status is correct here.
I'm still quite new to how these kind of conditions work for the GatewayAPI so I'm open to feedback and opinions on this one. Thanks!

Author: Okay, so in our case we need declare the rule as "invalid" and we will automatically add `PartiallyInvalid` by default today?

Also, when a route is `PartiallyInvalid` what is the outcome?
For example, if I have a valid route with a valid `AuthenticationFilter`, and I try to add another `AuthenticationFilter`, which would be invalid. Do we keep the pervious working state or does the route rule (`location`) get removed from NGINX?

  Type *TokenType `json:"type,omitempty"`

  // KeyCache is the cache duration for keys.
  // Configures auth_jwt_key_cache directive.
  // https://nginx.org/en/docs/http/ngx_http_auth_jwt_module.html#auth_jwt_key_cache
  // Example: "auth_jwt_key_cache 10m".
  //
  // +optional
  KeyCache *v1alpha1.Duration `json:"keyCache,omitempty"`

  // OnFailure customizes the 401 response for failed authentication.
  //
  // +optional
  OnFailure *AuthFailureResponse `json:"onFailure,omitempty"`
}

// FileKeySource specifies local JWKS key configuration.
type FileKeySource struct {
  // SecretRef references a Secret containing the JWKS.
  SecretRef LocalObjectReferenceWithKey `json:"secretRef,omitempty"`

  // KeyCache is the cache duration for keys.
  // Configures `auth_jwt_key_cache` directive.
  // https://nginx.org/en/docs/http/ngx_http_auth_jwt_module.html#auth_jwt_key_cache
  // Example: "auth_jwt_key_cache 10m;".
  //
  // +optional
  KeyCache *v1alpha1.Duration `json:"keyCache,omitempty"`
}

// LocalObjectReferenceWithKey sepcifies as local kubernetes object 
// with required `key` field to extract data.
type LocalObjectReferenceWithKey struct {
     v1.LocalObjectReference

>      Key string

Author: I'm not sure we should do that.
It feels like it would be safer to ensure the user is very explicit about which secret + key they want access to.
It's also not something that I've seen have a default in the past. There is a good chance that could cause confusion to the user.

>      Key string

Reviewer 2: I'm curious why you think it's safer to tell us the key? There are certain Secret types (like TLS) where the keys are defaulted and users don't have to tell us what they are, we assume because we've told them the type of Secret that it needs to be.

>      Key string

Author: That's a fair point, I didn't consider it like that. In my mind, it avoids a scenario where we define a default key like `htpaswd` which matches to a user shouldn't try to access. I might be over thinking the security implications of that.

>      Key string

Reviewer 2: I don't see an issue with having a default value for this.

>      Key string

Author: > conditional default in the code. If they're using a Basic Auth policy, the key is htpasswd, etc.
And we document this in the CRD (and our docs), and they can specify the key if it's something else.

That feels like quite a bit of overhead to allow a default.
Obviously we would want to document that, but I could see some time in the future trying to explain code like that to someone new to team who might get confused why and how it behaves that way. It doesn't feel like that really makes the user experience much smoother either for the effort involved.

### Golang API

Below is the Golang API for the `AuthenticationFilter` API:

```go
package v1alpha1

import (
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
  "github.com/nginx/nginx-gateway-fabric/v2/apis/v1alpha1"
)

// +genclient
// +kubebuilder:object:root=true
// +kubebuilder:storageversion
// +kubebuilder:subresource:status
// +kubebuilder:resource:categories=nginx-gateway-fabric,shortName=authfilter;authenticationfilter
// +kubebuilder:printcolumn:name="Age",type=date,JSONPath=`.metadata.creationTimestamp`

// AuthenticationFilter configures request authentication (Basic or JWT) and is
// referenced by HTTPRoute filters via ExtensionRef.
type AuthenticationFilter struct {
  metav1.TypeMeta   `json:",inline"`
  metav1.ObjectMeta `json:"metadata,omitempty"`

  // Spec defines the desired state of the AuthenticationFilter.
  Spec AuthenticationFilterSpec `json:"spec"`

  // Status defines the state of the AuthenticationFilter, following the same
  // pattern as SnippetsFilter: per-controller conditions with an Accepted condition.
  //
  // +optional
  Status AuthenticationFilterStatus `json:"status,omitempty"`
}

// +kubebuilder:object:root=true

> // AuthenticationFilterList contains a list of AuthenticationFilter.

Reviewer 3: ```suggestion
// AuthenticationFilterList contains a list of AuthenticationFilters.
```

type AuthenticationFilterList struct {

> 	metav1.TypeMeta `json:",inline"`

Reviewer 3: why do we need it here as it is already in the AuthenticationFilter struct line 82?

// +kubebuilder:validation:XValidation:message="when spec.basic is set, type must be 'Basic'",rule="self.basic != null ? self.type == 'Basic' : true"
// +kubebuilder:validation:XValidation:message="when spec.jwt is set, type must be 'JWT'",rule="self.jwt != null ? self.type == 'JWT' : true"
type AuthenticationFilterSpec struct {
  // Type selects the authentication mechanism.
  Type AuthType `json:"type"`

  // Basic configures HTTP Basic Authentication.
  // Required when Type == Basic.
  //
  // +optional
  Basic *BasicAuth `json:"basic,omitempty"`

  // JWT configures JSON Web Token authentication (NGINX Plus).
  // Required when Type == JWT.
  //
  // +optional
  JWT *JWTAuth `json:"jwt,omitempty"`
}

// AuthType defines the authentication mechanism.
// +kubebuilder:validation:Enum=Basic;JWT
type AuthType string

const (
  AuthTypeBasic AuthType = "Basic"
  AuthTypeJWT   AuthType = "JWT"
)

// BasicAuth configures HTTP Basic Authentication.
type BasicAuth struct {
  // SecretRef allows referencing a Secret in the same namespace

>   SecretRef LocalObjectReferenceWithKey `json:"secretRef,omitempty"`

Reviewer 2: If we're never going to support ConfigMaps for this (do we really need to?), then we probably don't actually need to use a LocalObjectReference that specifies the Kind. We could support just Name, Namespace, Key. But we should only do this if we know for sure, because otherwise it'll make our API a little messier in the future.

Author: I think we should just keep it as `SecretRef`.
If gives us the flexibility of eventually supporting ConfigMap, even if we never choose to.
It also has the added benefit of being familiar to users and it's very clear what it's trying to accomplish.

Author: @sjberman After going back and forth, it feels like it would be more secure to stick with just secret ref. The main benefit that ConfigMap offers is less overhead in encoding and decoding and it would be faster when dealing with less sensitive data data. As I looked over it, we mainly deal with usernames/passwords and JWKs in these scenarios, which I feel should always be encoded.

  // the condition is true.
  AuthenticationFilterConditionReasonAccepted AuthenticationFilterConditionReason = "Accepted"

  // AuthenticationFilterConditionReasonInvalid is used with the Accepted condition type when
  // the filter is invalid.
  AuthenticationFilterConditionReasonInvalid AuthenticationFilterConditionReason = "Invalid"
)
```

### Example spec for Basic Auth

```yaml
apiVersion: gateway.nginx.org/v1alpha1
kind: AuthenticationFilter
metadata:
  name: basic-auth
spec:
  type: Basic
  basic:
    secretRef:
      name: basic-auth-users   # Secret containing htpasswd data
      key: htpasswd
    realm: "Restricted"        # Optional. Helps with logging
    onFailure:                 # Optional. These setting may be defaults.
      statusCode: 401
      scheme: Basic
```

In the case of Basic Auth, the deployed Secret and HTTPRoute may look like this:

#### Secret referenced by filter

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: basic-auth-users
type: Opaque
stringData:
  htpasswd: |
    admin:$apr1$ZxY12345$abcdefghijklmnopqrstuvwx/
    user:$apr1$AbC98765$mnopqrstuvwxyzabcdefghiJKL/

> ```

Reviewer 3: do we have a structure describing this yaml for Secret?

#### HTTPRoute that will reference this filter

```yaml
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name: api-basic
spec:
  parentRefs:
  - name: gateway
  hostnames:
  - api.example.com
  rules:
  - matches:
    - path:
        type: PathPrefix
        value: /v2
    filters:
    - type: ExtensionRef
      extensionRef:
        group: gateway.nginx.org
        kind: AuthenticationFilter
        name: basic-auth
    backendRefs:
    - name: backend
      port: 80
```

#### Generated NGINX config

```nginx
http {
    upstream backend_default {
        server 10.0.0.10:80;
        server 10.0.0.11:80;
    }

    server {
        listen 80;
        server_name api.example.com;

        location /v2 {
            # Injected by BasicAuthFilter "basic-auth"
            auth_basic "Restricted";

>             auth_basic_user_file /etc/nginx/secrets/basic-auth-users/htpasswd;

Author: I can add that as a comment in the code

            # Optional: customize failure per filter onFailure
            # Ensures a consistent body and explicit WWW-Authenticate header
            error_page 401 = @basic_auth_failure;

            # Optional: do not forward client Authorization header to upstream
            proxy_set_header Authorization "";

            # NGF standard proxy headers
            proxy_set_header Host $host;
            proxy_set_header X-Real-IP $remote_addr;
            proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
            proxy_set_header X-Forwarded-Proto $scheme;

            # Pass traffic to upstream
            proxy_pass http://backend_default;
        }

        # Internal location for custom 401 response
        location @basic_auth_failure {
            add_header WWW-Authenticate 'Basic realm="Restricted"' always;
            return 401 'Unauthorized';

>         }

Author: There are two kinds of locations. The one in the NGINX docs is the location that the user would access.
For `location @basic_auth_failure`, this is referred to as a Named Location. It's called by the `error_page 401 = @basic_auth_failure;` directive. This will only get called if there is an authentication failure.

// +kubebuilder:validation:XValidation:message="when spec.basic is set, type must be 'Basic'",rule="self.basic != null ? self.type == 'Basic' : true"
// +kubebuilder:validation:XValidation:message="when spec.jwt is set, type must be 'JWT'",rule="self.jwt != null ? self.type == 'JWT' : true"
type AuthenticationFilterSpec struct {
	// Type selects the authentication mechanism.
	Type AuthType `json:"type"`

	// Basic configures HTTP Basic Authentication.
	//
	// +optional
	Basic *BasicAuth `json:"basic,omitempty"`

	// JWT configures JSON Web Token authentication (NGINX Plus).
	//
	// +optional
	JWT *JWTAuth `json:"jwt,omitempty"`
}

// AuthType defines the authentication mechanism.
// +kubebuilder:validation:Enum=Basic;JWT
type AuthType string

const (
	AuthTypeBasic AuthType = "Basic"
	AuthTypeJWT   AuthType = "JWT"
)

// BasicAuth configures HTTP Basic Authentication.
type BasicAuth struct {
	// SecretRef allows referencing a Secret in the same or different namespace.
	// When namespace is set and differs from the filter's namespace, a ReferenceGrant in the target namespace is required.
	//
	// +optional
	SecretRef *NamespacedSecretKeyReference `json:"secretRef,omitempty"`

> 	// Realm used by NGINX `auth_basic`.

Reviewer: ```suggestion
// Realm used by NGINX `auth_basic`.
```

// +kubebuilder:validation:XValidation:message="when remote is set, mode must be 'Remote'",rule="self.remote != null ? self.mode == 'Remote' : true"
type JWTAuth struct {
	// Realm used by NGINX `auth_jwt` directive.
  // Configures "realm="<realm_value>" in WWW-Authenticate header in error page location.
	//
	// +optional
  // +kubebuilder:default="Restricted"
	Realm *string `json:"realm,omitempty"`

	// Mode selects how JWT keys are provided: local file or remote JWKS.
	// Default: File.
	//
  // +kubebuilder:default=File"
	Mode JWTKeyMode `json:"mode,omitempty"`

	// File specifies local JWKS configuration (Secret or ConfigMap, mount path, file name).
	// Required when Mode == File. Exactly one of ConfigMapRef or SecretRef must be set.
	//
	// +optional
	File *JWTFileKeySource `json:"file,omitempty"`

	// Remote specifies remote JWKS configuration.
	// Required when Mode == Remote.
	//
	// +optional
	Remote *JWTRemoteKeySource `json:"remote,omitempty"`

	// Leeway is the acceptable clock skew for exp/nbf checks.
  // Configures `auth_jwt_leeway` directive.
	// Example: "auth_jwt_leeway 60s".
	//
	// +optional

>   // +kubebuilder:default=60s

Reviewer: these comment lines need to be aligned,

why are we defaulting it to 60s? nginx spec?

Author: It's strange, these aren't showing up as miss-aligned in my IDE.

Author: Okay, these should be fixed now. It was a mixed issue of some comments and fields being indented differently.

  //      "https://issuer.example1.com" 1;
  //      default 0;
  //  }
  //  map $jwt_claim_aud $valid_jwt_aud {
  //      "api" 1;
  //      "cli" 1;
  //      default 0;
  //  }
	//
	// +optional
	Require *JWTRequiredClaims `json:"require,omitempty"`

	// TokenSource defines where the client presents the token.
	// Defaults to Authorization header only.
	//
	// +optional
	TokenSource *JWTTokenSource `json:"tokenSource,omitempty"`

	// Propagation controls identity header propagation to upstream and header stripping.
	//
	// +optional
	Propagation *JWTPropagation `json:"propagation,omitempty"`
}

// JWTKeyMode selects where JWT keys come from.
type JWTKeyMode string

const (
	JWTKeyModeFile   JWTKeyMode = "File"
	JWTKeyModeRemote JWTKeyMode = "Remote"
)

// JWTFileKeySource specifies local JWKS key configuration.
// +kubebuilder:validation:XValidation:message="exactly one of configMapRef or secretRef must be set",rule="(self.configMapRef == null) != (self.secretRef == null)"
type JWTFileKeySource struct {
	// ConfigMapRef references a ConfigMap containing the JWKS.
	// Exactly one of ConfigMapRef or SecretRef must be set.
	//
	// +optional

> 	ConfigMapRef *NamespacedObjectReference `json:"configMapRef,omitempty"`

Author: That's true. It may be easier to just support Secrets. It would be more secure by default too.
I added this mostly to give users flexibility. After you asked this, I looked around for some use cases where someone might choose ConfigMapRef over SecretRef.

I found these reasons why a user might choose ConfigMapRef over SecretRef for JWKS:
- Non-sensitive data: Verification keys are public. Storing them in a Secret implies sensitivity that doesn’t exist unless you’re handling private/decryption keys (e.g., for auth_jwt_type=encrypted or nested). For public-only JWKS, ConfigMap is appropriate.
- Authoring convenience: ConfigMap stores raw JSON under data, while Secret requires base64-encoding. This avoids encoding errors and makes it simpler to read, diff, lint, and template the JWKS.
- GitOps friendliness: Public configuration can be safely committed to source control and templated with Helm/Kustomize. Secrets generally shouldn’t be committed to VCS.
- Operational visibility: Raw JSON is easier to inspect with kubectl for debugging and audits. No decoding step is needed to verify content.
- RBAC/Policy simplification: Many clusters restrict Secret access more tightly. If your controller/policy allows reading ConfigMaps more readily than Secrets, using ConfigMap avoids broader secret-read permissions.
- Avoiding unnecessary “secret” handling: Secrets may be encrypted at rest and trigger compliance pipelines and secret scanners. ConfigMaps avoid that overhead for public material.

> 	ConfigMapRef *NamespacedObjectReference `json:"configMapRef,omitempty"`

Reviewer 2: I think for a first pass, we just support Secret. And I think we can use a more generic object reference than SecretRef and ConfigmapRef. That way if we add ConfigMap support, we can reuse the existing object reference.

> 	ConfigMapRef *NamespacedObjectReference `json:"configMapRef,omitempty"`

Author: Good idea. I'll see what change might look like. Thanks @sjberman 😄

	// SecretRef references a Secret containing the JWKS (with optional key).
	// Exactly one of ConfigMapRef or SecretRef must be set.
	//
	// +optional
	SecretRef *NamespacedSecretKeyReference `json:"secretRef,omitempty"`

> 	// MountPath is the path where NGF will mount the data into the NGINX container.

Author: That's a good point. It's probably more secure and less error prone for us to default this directory.
As I think about that, we will want to ensure we can write to this directory when readOnlyRootfileSystem is enabled.

  // Used in `auth_jwt_key_file` directive.
	// Example: "/etc/nginx/keys".
	MountPath string `json:"mountPath"`

> 	// FileName is the file name of the JWKS within the mount path.

Author: This has been changed now

	//
	// +optional
  // +kubebuilder:default=true
	Header *bool `json:"header,omitempty"`

	// Read token from a cookie. Default: false.
	//
	// +optional
  // +kubebuilder:default=false
	Cookie *bool `json:"cookie,omitempty"`

	// CookieName when Cookie is true. Example: "access_token".
	//
	// +optional
  // +kubebuilder:default=access_token
	CookieName *string `json:"cookieName,omitempty"`

	// Read token from query string. Default: false.
	//
	// +optional
  // +kubebuilder:default=false
	Query *bool `json:"query,omitempty"`

	// QueryParam when Query is true. Example: "access_token".
	//
	// +optional
  // +kubebuilder:default=access_token
	QueryParam *string `json:"queryParam,omitempty"`
}

// JWTPropagation controls identity header propagation and header stripping.
type JWTPropagation struct {
	// AddIdentityHeaders defines headers to add on success with values
	// typically derived from jwt_claim_* variables.
	//
	// +optional
	AddIdentityHeaders []HeaderValue `json:"addIdentityHeaders,omitempty"`

	// StripAuthorization removes the incoming Authorization header before proxying.
	//
	// +optional
	StripAuthorization *bool `json:"stripAuthorization,omitempty"`
}

> // HeaderValue defines a header name and a value (may reference NGINX variables).

Author: Resolving this one as we've covered it in the Stretch Goals section

- Design for OIDC Auth
- An Auth filter for TCP and UDP routes
- Design for integration with [ExternalAuth in the Gateway API](https://gateway-api.sigs.k8s.io/geps/gep-1494/)

## Introduction

This document focuses expliclty on Authentication (AuthN) and not Authorization (AuthZ). Authentication (AuthN) defines the verification of identiy. It asks the question, "Who are you?". This is different from Authorization (AuthZ), which preceeds Authentication. It asks the question, "What are you allowed to do".

This document also focus on HTTP Basic Authentication and JWT Authentication. Other authentication methods such as OpenID Connect (OIDC) are mentioned, but are not part of the CRD design. These will be covered in future design and implementation tasks.

## Use Cases

- As an Application Developer, I want to secure access to my APIs and Backend Applications.

> - As an Application Developer, I want to enforce authenticaiton on specific routes and matches.

Reviewer 4: ```suggestion
- As an Application Developer, I want to enforce authentication on specific routes and matches.
```

| **Authentication Method**     | **OSS**      | **Plus** | **NGINX Module**                | **Details**                                                        |
|-------------------------------|--------------|----------------|----------------------------------|--------------------------------------------------------------------|
| **HTTP Basic Authentication** | ✅           | ✅             | [ngx_http_auth_basic](https://nginx.org/en/docs/http/ngx_http_auth_basic_module.html) | Requires a username and password sent in an HTTP header.           |
| **JWT (JSON Web Token)**     | ❌           | ✅             | [ngx_http_auth_jwt_module](https://nginx.org/en/docs/http/ngx_http_auth_jwt_module.html) | Tokens are used for stateless authentication between client and server. |
| **OpenID Connect**            | ❌           | ✅             | [ngx_http_oidc_module](https://nginx.org/en/docs/http/ngx_http_oidc_module.html)| Allows authentication through third-party providers like Google.   |

## API, Customer Driven Interfaces, and User Experience

This portion of the proposal will cover API design and interaction experience for use of Basic Auth and JWT.

> This portioan also contains:

Reviewer 4: ```suggestion
This portion also contains:
```

1. The Golang API
2. Example spec for Basic Auth

>     - Example HTTPRoutes and NINGX configuration

Reviewer 4: ```suggestion
    - Example HTTPRoutes and NGINX configuration
```

3. Example spec for JWT Auth
    - Example HTTPRoutes
    - Examples for Local & Remote JWKS configration
    - Example NINGX configuration for both Local & Remote JWKS

>     - Example of additioanl optional fields

Reviewer: this still needs to be addressed

	//
	// +optional
	OnFailure *AuthFailureResponse `json:"onFailure,omitempty"`
}

// JWTAuth configures JWT-based authentication (NGINX Plus).
// +kubebuilder:validation:XValidation:message="mode 'File' requires file set and remote unset",rule="self.mode == 'File' ? self.file != null && self.remote == null : true"
// +kubebuilder:validation:XValidation:message="mode 'Remote' requires remote set and file unset",rule="self.mode == 'Remote' ? self.remote != null && self.file == null : true"
// +kubebuilder:validation:XValidation:message="when file is set, mode must be 'File'",rule="self.file != null ? self.mode == 'File' : true"
// +kubebuilder:validation:XValidation:message="when remote is set, mode must be 'Remote'",rule="self.remote != null ? self.mode == 'Remote' : true"
type JWTAuth struct {
	// Realm used by NGINX `auth_jwt` directive.
  // Configures "realm="<realm_value>" in WWW-Authenticate header in error page location.
	//
	// +optional
  // +kubebuilder:default="Restricted"
	Realm *string `json:"realm,omitempty"`

	// Mode selects how JWT keys are provided: local file or remote JWKS.
	// Default: File.
	//
	// +optional

>   // +kubebuilder:default=File

Reviewer 4: ```suggestion
    // +kubebuilder:default=File
```

There's a few other places where the indentation needs to be fixed

// +kubebuilder:validation:XValidation:message="exactly one of configMapRef or secretRef must be set",rule="(self.configMapRef == null) != (self.secretRef == null)"
type JWTFileKeySource struct {
	// ConfigMapRef references a ConfigMap containing the JWKS.
	// Exactly one of ConfigMapRef or SecretRef must be set.
	//
	// +optional
	ConfigMapRef *NamespacedObjectReference `json:"configMapRef,omitempty"`

	// SecretRef references a Secret containing the JWKS (with optional key).
	// Exactly one of ConfigMapRef or SecretRef must be set.
	//
	// +optional
	SecretRef *NamespacedSecretKeyReference `json:"secretRef,omitempty"`

	// MountPath is the path where NGF will mount the data into the NGINX container.
  // Used in `auth_jwt_key_file` directive.
	// Example: "/etc/nginx/keys".
	MountPath string `json:"mountPath"`

	// FileName is the file name of the JWKS within the mount path.
  // Used in `auth_jwt_key_file` directive.
	// Example: "jwks.json".
	FileName string `json:"fileName"`

	// KeyCache is the cache duration for keys.
  // Configures `auth_jwt_key_cache` directive
	// Example: "auth_jwt_key_cache 10m;".
	//
	// +optional

> 	KeyCache *string `json:"keyCache,omitempty"`

Reviewer 2: It allows us to perform validation on the format of it.

	// typically derived from jwt_claim_* variables.
	//
	// +optional
	AddIdentityHeaders []HeaderValue `json:"addIdentityHeaders,omitempty"`

	// StripAuthorization removes the incoming Authorization header before proxying.
	//
	// +optional
	StripAuthorization *bool `json:"stripAuthorization,omitempty"`
}

// HeaderValue defines a header name and a value (may reference NGINX variables).
type HeaderValue struct {
	Name      string `json:"name"`
	ValueFrom string `json:"valueFrom"`
}

// AuthScheme enumerates supported WWW-Authenticate schemes.
type AuthScheme string

const (
    AuthSchemeBasic  AuthScheme = "Basic"
    AuthSchemeBearer AuthScheme = "Bearer"
)

// AuthFailureBodyPolicy controls the failure response body behavior.
type AuthFailureBodyPolicy string

const (
    AuthFailureBodyPolicyUnauthorized AuthFailureBodyPolicy = "Unauthorized"
    AuthFailureBodyPolicyForbidden AuthFailureBodyPolicy = "Forbidden"
    AuthFailureBodyPolicyEmpty   AuthFailureBodyPolicy = "Empty"
)

// AuthFailureResponse customizes 401/403 failures.
type AuthFailureResponse struct {
    // Allowed: 401, 403.
    // Default: 401.
    //
    // +optional
    // +kubebuilder:default=401

>     // +kubebuilder:validation:XValidation:message="statusCode must be 401 or 403",rule="self == null || self in [401, 403]"

Author: Not sure. I'll see if it's possible though.

>     // +kubebuilder:validation:XValidation:message="statusCode must be 401 or 403",rule="self == null || self in [401, 403]"

Reviewer: i think you should be able to since they resolve as strings?

Below is an eample of an **invalid** HTTPRoute that references multiple `AuthenticationFilter` resources in a single rule

```yaml
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name: invalid-httproute
spec:
  parentRefs:
  - name: gateway
  hostnames:
  - api.example.com
  rules:
  - matches:
    - path:
        type: PathPrefix
        value: /api
    filters:
    - type: ExtensionRef
      extensionRef:
        group: gateway.nginx.org
        kind: AuthenticationFilter
        name: basic-auth
    - type: ExtensionRef
      extensionRef:
        group: gateway.nginx.org
        kind: AuthenticationFilter
        name: jwt-auth
    backendRefs:
    - name: backend
      port: 80
```

## Testing

- Unit tests
- Functional tests to validate behavioural scenarios when referncing filters in different combinations. The details of these tests are out of scope for this document.

## Security Considerations

### Basic Auth and Local JWKS

Basic Auth sends credentials in an Authorization header that is base64-encoded.

> JWT Auth requires users to provided a bearer token through the Authroization header.

Reviewer 4: ```suggestion
JWT Auth requires users to provide a bearer token through the Authorization header.
```

  to:
  - group: ""   # core API group
    kind: Secret
    name: basic-auth-users
```

AuthenticationFilter referencing the cross-namespace Secret
```yaml
apiVersion: gateway.nginx.org/v1alpha1
kind: AuthenticationFilter
metadata:
  name: basic-auth
  namespace: app-ns
spec:
  type: Basic
  basic:
    secretRef:
      namespace: security-ns
      name: basic-auth-users
      key: htpasswd
    realm: "Restricted"
```

Example: Grant JWT file-based JWKS in keys-ns to filter in app-ns
```yaml
apiVersion: gateway.networking.k8s.io/v1
kind: ReferenceGrant
metadata:
  name: allow-jwks-configmap
  namespace: keys-ns
spec:
  from:
  - group: gateway.nginx.org
    kind: AuthenticationFilter
    namespace: app-ns
  to:
  - group: ""    # core API group
    kind: ConfigMap
    name: jwt-keys
```

AuthenticationFilter referencing cross-namespace JWKS ConfigMap
```yaml
apiVersion: gateway.nginx.org/v1alpha1
kind: AuthenticationFilter
metadata:
  name: jwt-auth
  namespace: app-ns
spec:
  type: JWT
  jwt:
    mode: File
    file:
      configMapRef:
        namespace: keys-ns
        name: jwt-keys
      mountPath: /etc/nginx/keys
      fileName: jwks.json
```

### Remote JWKS

> Proxy cache TTL should be configurable and set to a resonable default, reducing periods of stale cached JWKs.

Reviewer 4: ```suggestion
Proxy cache TTL should be configurable and set to a reasonable default, reducing periods of stale cached JWKs.
```

### Key rotation

Users sholud be advised to regularly rotate their JWKS keys in cases where they chose to reference a local JWKS via a `secrefRef` or `configMapRef`

### Auth failure behaviour

3xx response codes should not be allowed and AuthenticationFilter.onFailure must not support redirect targets. This is to prevent to prevent open-redirect abuse.

401 and 403 should be the only allowable auth failure codes.

### Auth failure default headers

Below are a list of default defensive headers for authentication failure reponses.
We may choose to include these headers by default for improved robustness in auth falure responses.

```nginx
add_header Content-Type "text/plain; charset=utf-8" always;
add_header X-Content-Type-Options "nosniff" always;
add_header Cache-Control "no-store" always;

> add_header Pragma "no-cache" always;

Reviewer 4: I had to look this up, I had never seen it before. It's been superseded by Cache-Control, I don't think we would need it, it's legacy from HTTP/1.0

```

Detailed header breakdown:

- Content-Type: "text/plain; charset=utf-8"

>   - This header explicitly set the body as plan text. This prevents browsers from treating the response as HTML or JavaScript, and is effective at mitigating Cross-side scrpting (XSS) through error pages

Reviewer 2: ```suggestion
  - This header explicitly set the body as plain text. This prevents browsers from treating the response as HTML or JavaScript, and is effective at mitigating Cross-side scrpting (XSS) through error pages
```

- Cache-Control: "no-store"
  - This header informs browsers and proxies not to cache the response. Avoids sensitive, auth-related content, from being being stored and served later to unintended recipients.

- Pragma: "no-cache"
  - This header is commonly paired with `Cache-Control: "no-store"` for broad coverage. It acts as an additional signal for older intermediaries that do not honor Cache-Control.

### Validation

When referencing an `AuthenticationFilter` in either a HTTPRoute or GRPCRoute, it is important that we ensure all configurable fields are validated, and that the resulting NGINX configuration is correct and secure

All fields in the `AuthenticationFilter` will be validated with Open API Schema.
We should also include [CEL](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules) validation where required.

We should validated that only one `AuthenticationFilter` is referenced per-rule. Multiple references to an `AuthenticationFilter` in a single rule should result in an `Invalid` HTTPRoute/GRPCRoute, and the resource should be `Rejected`.

> an `AuthenticationFilter` that sets a `onFailure.statusCode` to anything other than `401` or `403` should be rejected. This relates to the "Auth failure behaviour" section in the Security Condierations section.

Reviewer 2: ```suggestion
An `AuthenticationFilter` that sets a `onFailure.statusCode` to anything other than `401` or `403` should be rejected. This relates to the "Auth failure behaviour" section in the Security Considerations section.
```

## Alternatives

The Gateway API defines a means to standardise authentication through use of the [HTTPExternalAuthFilter](https://gateway-api.sigs.k8s.io/reference/spec/#httpexternalauthfilter) available in the HTTPRoute specification.

This allows users to reference an external authentication services, such as Keycloak, to handle the authentication requests.
While this API is available in the experimental channel, it is subject to change.

> Our decision to go forward with our own `AuthenticationFilter` was to ensure we could quckly provide authenticaiton to our users while allowing us to closley monitor progress of the ExternalAuthFilter.

Author: @sjberman that's not entirely true. If we went forward with ExternalAuth filter, we could build out an external auth service that uses NGINX. It wouldn't prevent us from supplying those auth methods.

> Our decision to go forward with our own `AuthenticationFilter` was to ensure we could quckly provide authenticaiton to our users while allowing us to closley monitor progress of the ExternalAuthFilter.

Author: @sjberman do you think this statement needs to be changed in any way?

It is certainly possible for us to provide an External Authentication Services that leverages NGINX and is something we can further investigate as the API progresses.

## Additional considerations

### Documenting filter behavour

> In regards to documentation of filter behavour with the `AuthenticationFilter`, the Gateway API documentation on filters states the following:

Reviewer 4: ```suggestion
In regards to documentation of filter behaviour with the `AuthenticationFilter`, the Gateway API documentation on filters states the following:
```

  // Also configures "realm="<realm_value>" in WWW-Authenticate header in error page location.
  //
  // +optional

>   // +kubebuilder:default=""

Author: Not sure why this hasn't shown as changed in this comment.
I did check, and the default realm values for BaiscAuth and JWTAuth are set to `""` in the recent changes.
For Baisc Auth: https://github.com/nginx/nginx-gateway-fabric/pull/4235/files#diff-f8d64ed0ce55db8bfa28f3bbb66023efd2604c2c932e599832fda3de21c4803eR157-R163

For JWT Auth: https://github.com/nginx/nginx-gateway-fabric/pull/4235/files#diff-f8d64ed0ce55db8bfa28f3bbb66023efd2604c2c932e599832fda3de21c4803eR186-R192

// FileKeySource specifies local JWKS key configuration.
type FileKeySource struct {
  // SecretRef references a Secret containing the JWKS.

>   SecretRef LocalObjectReferenceWithKey `json:"secretRef,omitempty"`

Reviewer 2: Could be something like `keySourceRef` for this particular case.

  KeyCache *v1alpha1.Duration `json:"keyCache,omitempty"`
}

// LocalObjectReferenceWithKey specifies a local Kubernetes object
// with a required `key` field to extract data.
// +kubebuilder:validation:XValidation:message="name must be set",rule="self.name != ''"
type LocalObjectReferenceWithKey struct {
    // Inline the core LocalObjectReference so `name` sits at the same JSON level
    // (optional, but avoids nesting)
    v1.LocalObjectReference `json:",inline"`

    // Key must be provided and non-empty.
    // This makes the field required in the OpenAPI schema.
    // +kubebuilder:validation:MinLength=1

>     // +kubebuilder:validation:XValidation:rule="self != ''",message="key must be non-empty"

Reviewer 2: CEL not needed here, adds nothing that the OpenAPI validation doesn't already do.

	Aud *string `json:"aud,omitempty"`
}

 // JWTTokenSourceMode selects where the JWT token is read from.
type JWTTokenSourceMode string

const (
	// Read from Authorization header (Bearer). Default.
	JWTTokenSourceModeHeader JWTTokenSourceMode = "Header"
	// Read from a cookie named tokenName.
	JWTTokenSourceModeCookie JWTTokenSourceMode = "Cookie"
	// Read from a query arg named tokenName.
	JWTTokenSourceModeQueryArg JWTTokenSourceMode = "QueryArg"
)

// JWTTokenSource specifies where tokens may be read from and the name when required.
type JWTTokenSource struct {
	// Mode selects the token source.

> 	// +kubebuilder:validation:Enum=Header;Cookie;QueryArg

Reviewer 2: this validation can be removed from here
